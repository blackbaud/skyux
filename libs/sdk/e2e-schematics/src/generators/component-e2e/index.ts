import { SchematicsAngularApplicationStyle } from '@angular/cli/lib/config/workspace-schema';
import { applicationGenerator } from '@nx/angular/generators';
import {
  ProjectConfiguration,
  Tree,
  formatFiles,
  generateFiles,
  getProjects,
  installPackagesTask,
  joinPathFragments,
  logger,
  readJson,
  readProjectConfiguration,
  removeDependenciesFromPackageJson,
  updateJson,
  updateProjectConfiguration,
} from '@nx/devkit';
import { addDependenciesToPackageJson } from '@nx/devkit/src/utils/package-json';
import { Linter } from '@nx/linter';
import { configurationGenerator } from '@nx/storybook';
import { moveGenerator } from '@nx/workspace';

import configurePercy from '../configure-percy';
import configureStorybook from '../configure-storybook';
import storiesGenerator from '../stories';

import { NormalizedSchema, Schema } from './schema';

const BASE_PATH = 'e2e';

function normalizeOptions(options: Partial<Schema>): NormalizedSchema {
  if (!options.name) {
    throw new Error(`Please provide the component library name`);
  }

  const parsedTags = (
    options.tags ? options.tags.split(',').map((s) => s.trim()) : []
  ).concat(['component-e2e']);

  return {
    name: options.name,
    prefix: 'app',
    storybookAppName: `${options.name}-storybook`,
    routing: true,
    strict: false,
    style: SchematicsAngularApplicationStyle.Scss,
    parsedTags,
    tags: parsedTags.join(','),
    ansiColor: options.ansiColor !== false,
  };
}

/**
 * Drop the BASE_PATH from the project name and re-sort the projects.
 */
function simplifyWorkspaceName(tree: Tree, projectName: string) {
  const projects = getProjects(tree);
  projects.forEach((projectConfig) => {
    let projectConfigJson = JSON.stringify(projectConfig).replace(
      new RegExp(`${BASE_PATH}-${projectName}`, 'g'),
      projectName
    );
    if (projectName.endsWith('-e2e')) {
      const withoutE2e = projectName.replace(/-e2e$/, '');
      projectConfigJson = projectConfigJson.replace(
        new RegExp(`${BASE_PATH}-${withoutE2e}`, 'g'),
        withoutE2e
      );
    }
    updateProjectConfiguration(
      tree,
      `${BASE_PATH}-${projectName}`,
      JSON.parse(projectConfigJson)
    );
  });
}

/**
 * Add the packages polyfills to the build and test targets.
 */
function addPackagesPolyfills(tree: Tree, projectName: string) {
  const polyfillsBuilders = [
    '@angular-devkit/build-angular:browser',
    '@angular-devkit/build-angular:karma',
    '@blackbaud-internal/skyux-angular-builders:browser',
    '@blackbaud-internal/skyux-angular-builders:karma',
  ];
  const projects = getProjects(tree);
  const projectConfig = projects.get(projectName);
  if (projectConfig) {
    let hasChanged = false;
    ['build', 'test'].forEach((target) => {
      if (
        polyfillsBuilders.includes(
          `${projectConfig.targets?.[target].executor}`
        ) &&
        projectConfig.targets?.[target].options.polyfills &&
        Array.isArray(projectConfig.targets[target].options.polyfills)
      ) {
        projectConfig.targets[target].options.polyfills.push(
          'libs/components/packages/src/polyfills.js'
        );
        hasChanged = true;
      }
    });
    if (hasChanged) {
      updateProjectConfiguration(tree, projectName, projectConfig);
    }
  }
}

/**
 * - Generates -storybook and -storybook-e2e projects for a component library if they don't already exist.
 * - Applies configuration to the projects.
 * - Generates stories and tests if there are demonstration components in the -storybook project.
 */
export default async function (tree: Tree, schema: Partial<Schema>) {
  const options = normalizeOptions(schema);

  let createProject = false;
  let moveProject = false;
  /* istanbul ignore next */
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  let projectConfig: ProjectConfiguration;
  let e2eProjectConfig: ProjectConfiguration;
  try {
    projectConfig = readProjectConfiguration(tree, options.storybookAppName);
    if (projectConfig.root === `apps/${options.storybookAppName}`) {
      moveProject = true;
      await moveGenerator(tree, {
        projectName: options.storybookAppName,
        destination: `${BASE_PATH}/${options.storybookAppName}`,
        updateImportPath: false,
        skipFormat: true,
      });
      simplifyWorkspaceName(tree, options.storybookAppName);
      projectConfig = readProjectConfiguration(tree, options.storybookAppName);
    }
    e2eProjectConfig = readProjectConfiguration(
      tree,
      `${options.storybookAppName}-e2e`
    );
    if (e2eProjectConfig.root === `apps/${options.storybookAppName}-e2e`) {
      moveProject = true;
      await moveGenerator(tree, {
        projectName: `${options.storybookAppName}-e2e`,
        destination: `${BASE_PATH}/${options.storybookAppName}-e2e`,
        updateImportPath: false,
        skipFormat: true,
      });
      simplifyWorkspaceName(tree, `${options.storybookAppName}-e2e`);
      e2eProjectConfig = readProjectConfiguration(
        tree,
        `${options.storybookAppName}-e2e`
      );
    }
    if (!moveProject) {
      (schema.ansiColor === false ? console.warn : logger.warn)(
        `The project "${options.storybookAppName}" already exists.`
      );
    }
  } catch (e) {
    createProject = true;
    await applicationGenerator(tree, {
      name: options.storybookAppName,
      tags: options.tags,
      style: options.style,
      routing: options.routing,
      strict: options.strict,
      prefix: options.prefix,
      directory: BASE_PATH,
      skipPackageJson: true,
      standaloneConfig: true,
      standalone: false,
    });
    simplifyWorkspaceName(tree, options.storybookAppName);
    simplifyWorkspaceName(tree, `${options.storybookAppName}-e2e`);
    addPackagesPolyfills(tree, options.storybookAppName);
    projectConfig = readProjectConfiguration(tree, options.storybookAppName);
    e2eProjectConfig = readProjectConfiguration(
      tree,
      `${options.storybookAppName}-e2e`
    );

    // Delete boilerplate files from the storybook project.
    let indexFile = tree.read(`${projectConfig.sourceRoot}/index.html`, 'utf8');

    // istanbul ignore if
    if (!indexFile) {
      indexFile = '';
    }

    indexFile = indexFile.replace(
      '<link rel="icon" type="image/x-icon" href="favicon.ico" />',
      ''
    );
    tree.write(`${projectConfig.sourceRoot}/index.html`, indexFile);
    [
      'favicon.ico',
      'assets/.gitkeep',
      'app/app.module.ts',
      'app/app.component.scss',
      'app/app.component.html',
      'app/app.component.spec.ts',
      'app/app.component.ts',
      'app/nx-welcome.component.ts',
    ].forEach((file) => tree.delete(`${projectConfig.sourceRoot}/${file}`));
    ['e2e/app.cy.ts', 'fixtures/example.json', 'support/app.po.ts'].forEach(
      (file) => tree.delete(`${e2eProjectConfig.sourceRoot}/${file}`)
    );
    // Create an empty app.
    generateFiles(
      tree,
      joinPathFragments(__dirname, 'files/app'),
      `${projectConfig.sourceRoot}/app`,
      {}
    );
    tree.write(`${e2eProjectConfig.sourceRoot}/e2e/.gitkeep`, ``);
  }

  /* istanbul ignore next */
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  if (
    createProject ||
    !(
      tree.isFile(`${projectConfig.root}/.storybook/main.js`) ||
      tree.isFile(`${projectConfig.root}/.storybook/main.ts`)
    )
  ) {
    await configurationGenerator(tree, {
      name: options.storybookAppName,
      uiFramework: '@storybook/angular',
      configureCypress: true,
      linter: Linter.EsLint,
    });
  } else if (!moveProject) {
    await storiesGenerator(tree, {
      project: options.storybookAppName,
      cypressProject: `${options.storybookAppName}-e2e`,
      generateCypressSpecs: true,
    });
  }
  if (!moveProject) {
    await configureStorybook(tree, { name: options.storybookAppName });
    await configurePercy(tree, { name: `${options.storybookAppName}-e2e` });
  }

  // @storybook/addon-essentials includes docs, which requires several other dependencies.
  // We install only the dependencies we need, and the storybook version is different from the one
  // that NX forces.
  const packageJson = readJson(tree, 'package.json');
  const storybookVersion =
    packageJson.devDependencies['@storybook/addon-toolbars'];
  if (storybookVersion) {
    // The `addDependenciesToPackageJson` will not set the right version if the version is lower than the version than what the `storiesGenerator`
    // installs (even if this version uses a caret). Remove the entries first so that the addition will go through with the right version.
    removeDependenciesFromPackageJson(
      tree,
      [],
      ['@storybook/angular', '@storybook/core-server']
    );
    addDependenciesToPackageJson(
      tree,
      {},
      {
        '@storybook/angular': storybookVersion,
        '@storybook/core-server': storybookVersion,
      }
    );
  }
  // Do not add explicit dependencies for @storybook/addon-essentials or webpack.
  removeDependenciesFromPackageJson(
    tree,
    [],
    ['@storybook/addon-essentials', 'html-webpack-plugin', 'webpack']
  );

  // Clean up duplicate entries in nx.json
  updateJson(tree, 'nx.json', (json) => {
    if (
      json.targetDefaults['build-storybook']?.inputs &&
      Array.isArray(json.targetDefaults['build-storybook'].inputs)
    ) {
      // Remove duplicate entries
      json.targetDefaults['build-storybook'].inputs = json.targetDefaults[
        'build-storybook'
      ].inputs.filter(
        (v: string, i: number, a: string[]) => a.indexOf(v) === i
      );
    }
    if (
      json.namedInputs?.production &&
      Array.isArray(json.namedInputs.production)
    ) {
      // Remove duplicate entries
      json.namedInputs.production = json.namedInputs.production.filter(
        (v: string, i: number, a: string[]) => a.indexOf(v) === i
      );
    }
    return json;
  });

  await formatFiles(tree);

  /* istanbul ignore next */
  return () => installPackagesTask(tree);
}
