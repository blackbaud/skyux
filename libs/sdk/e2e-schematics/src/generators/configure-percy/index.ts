import type { Tree } from '@nx/devkit';
import { formatFiles, generateFiles, joinPathFragments } from '@nx/devkit';
import { insertStatement } from '@nx/workspace/src/generators/utils/insert-statement';

import { getE2eProjects } from '../../utils';

import { Schema } from './schema';

/**
 * Disable video recording and import @percy/cypress into the e2e project, which enables the .percySnapshot method.
 */
export default async function (tree: Tree, schema: Schema) {
  const projects = getE2eProjects(tree, schema.name);
  projects.forEach((project) => {
    // Disable video recording
    generateFiles(
      tree,
      joinPathFragments(__dirname, 'files'),
      project.root,
      {}
    );
    const filePath = joinPathFragments(project.root, 'src/support/e2e.ts');
    const importPercyCypress = `import` + ` '@percy/cypress';`;
    if (!tree.isFile(filePath)) {
      tree.write(filePath, importPercyCypress);
    } else {
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const content = tree.read(filePath)!.toString();
      if (!content.includes(importPercyCypress)) {
        insertStatement(tree, filePath, importPercyCypress);
      }
    }
  });

  return formatFiles(tree);
}
