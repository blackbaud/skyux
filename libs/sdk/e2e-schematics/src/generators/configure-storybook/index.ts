import {
  ProjectConfiguration,
  TargetConfiguration,
  Tree,
  formatFiles,
  generateFiles,
  getWorkspacePath,
  joinPathFragments,
  logger,
  normalizePath,
  readProjectConfiguration,
  updateJson,
} from '@nrwl/devkit';
import { TsConfig } from '@nrwl/storybook/src/utils/utilities';

import { updateProjectConfiguration } from 'nx/src/generators/utils/project-configuration';
import { relative } from 'path';

import { getStorybookProjects } from '../../utils';

import { Schema } from './schema';

/**
 * Configure Storybook to use typescript. Set Storybook to run during e2e in development mode.
 */
export default async function (tree: Tree, schema: Schema) {
  const projects = getStorybookProjects(tree, schema.name);
  const workspacePath = getWorkspacePath(tree);

  // istanbul ignore if
  if (!workspacePath) {
    throw new Error(`Unable to determine workspace file path`);
  }

  // istanbul ignore next
  const errorLogger = schema.ansiColor === false ? console.error : logger.fatal;

  projects.forEach((project, projectName) => {
    let hasChanged = false;
    const targets = Object.keys(
      project.targets as Record<string, TargetConfiguration>
    );
    targets.forEach((target) => {
      project.targets = project.targets as Record<string, TargetConfiguration>;
      const targetConfig = project.targets[target] as TargetConfiguration;
      // Does the builder support styles?
      if (
        [
          '@angular-devkit/build-angular:browser',
          '@storybook/angular:build-storybook',
          '@storybook/angular:start-storybook',
        ].includes(targetConfig.executor)
      ) {
        // Add stylesheets to project.
        if (!targetConfig.options) {
          targetConfig.options = {
            styles: [],
          };
        }
        [
          'libs/components/theme/src/lib/styles/sky.scss',
          'libs/components/theme/src/lib/styles/themes/modern/styles.scss',
        ].forEach((stylesheet) => {
          if (!targetConfig.options.styles?.includes(stylesheet)) {
            hasChanged = true;
            if (targetConfig.options.styles) {
              targetConfig.options.styles.push(stylesheet);
            } else {
              targetConfig.options.styles = [stylesheet];
            }
          }
        });
      }
      // Drop the asset path.
      if (target === 'build' && targetConfig.options.assets) {
        hasChanged = true;
        delete targetConfig.options.assets;
      }
    });
    if (hasChanged) {
      updateProjectConfiguration(tree, projectName, project);
    }

    const e2eProjectName = `${projectName}-e2e`;
    let e2eProject: ProjectConfiguration | undefined = undefined;
    try {
      e2eProject = readProjectConfiguration(tree, e2eProjectName);
    } catch (e) {
      errorLogger(`Project "${e2eProjectName}" does not exist`);
    }
    if (
      e2eProject &&
      e2eProject.targets?.e2e &&
      e2eProject.targets.e2e.executor === '@nrwl/cypress:cypress'
    ) {
      let hasChanged = false;
      if (
        !e2eProject.targets.e2e.options.devServerTarget ||
        !e2eProject.targets.e2e.options.baseUrl
      ) {
        hasChanged = true;
        // During development, run Storybook and Cypress.
        e2eProject.targets.e2e.options = {
          ...e2eProject.targets.e2e.options,
          devServerTarget: `${projectName}:storybook`,
          baseUrl: `http://localhost:4400`,
        };
      }
      if (hasChanged) {
        updateProjectConfiguration(tree, e2eProjectName, e2eProject);
      }
    } else {
      errorLogger(
        `Project "${e2eProjectName}" does not have an e2e target with @nrwl/cypress:cypress`
      );
    }

    const projectRoot = project.root;
    const relativeToRoot = normalizePath(
      relative(`/${projectRoot}/.storybook`, `/`)
    );

    const tsconfigFile = `${projectRoot}/.storybook/tsconfig.json`;
    const tsconfigAppFile = `${projectRoot}/tsconfig.app.json`;
    if (!tree.isFile(tsconfigFile)) {
      tree.write(
        `${projectRoot}/.storybook/tsconfig.json`,
        JSON.stringify({
          extends: '../tsconfig.json',
          compilerOptions: {
            emitDecoratorMetadata: true,
          },
          exclude: ['../**/*.spec.ts', 'jest.config.ts'],
          include: ['../src/**/*', './*'],
        })
      );
    }
    updateJson(tree, tsconfigFile, (tsconfig: TsConfig) => {
      if (!tsconfig.include) {
        tsconfig.include = [];
      }
      if (!tsconfig.include.includes('./*')) {
        tsconfig.include.push('./*');
      }
      if (tsconfig.exclude && !tsconfig.exclude.includes('jest.config.ts')) {
        tsconfig.exclude.push('jest.config.ts');
      }
      return tsconfig;
    });
    if (tree.isFile(tsconfigAppFile)) {
      updateJson(tree, tsconfigAppFile, (tsconfig: TsConfig) => {
        if (tsconfig.exclude && !tsconfig.exclude.includes('jest.config.ts')) {
          tsconfig.exclude.push('jest.config.ts');
        }
        return tsconfig;
      });
    }

    // Remove default files.
    ['preview', 'main'].forEach((file) => {
      tree.delete(`${projectRoot}/.storybook/${file}.js`);
    });

    if (!tree.isFile(`${projectRoot}/.storybook/preview.ts`)) {
      generateFiles(
        tree,
        joinPathFragments(__dirname, `./files`),
        `${projectRoot}/.storybook/`,
        {
          ...schema,
          relativeToRoot,
        }
      );
    }

    if (!tree.isFile(`${projectRoot}/.storybook/manager.ts`)) {
      tree.write(
        `${projectRoot}/.storybook/manager.ts`,
        `export * from '${relativeToRoot}/.storybook/manager';`
      );
    }
  });

  await formatFiles(tree);
}
