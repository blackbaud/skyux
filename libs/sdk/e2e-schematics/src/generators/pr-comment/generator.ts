import { Tree, generateFiles, joinPathFragments } from '@nx/devkit';

import { PrCommentGeneratorOptions, PrCommentGeneratorSchema } from './schema';

function normalizeOptions(
  tree: Tree,
  options: PrCommentGeneratorSchema
): PrCommentGeneratorOptions {
  const pr = options.pr;
  const storybooks = options.storybooks
    .split(',')
    .map((s) => s.trim())
    .filter((s) => s.length > 0);
  const apps = (options.apps || '')
    .split(',')
    .map((s) => s.trim())
    .filter((s) => s.length > 0);
  if (!pr || isNaN(parseInt(`${pr}`, 10))) {
    throw new Error('Invalid PR number');
  }
  return {
    pr,
    storybooks,
    apps,
  };
}

export default async function prComment(
  tree: Tree,
  options: PrCommentGeneratorSchema
) {
  const normalizedOptions = normalizeOptions(tree, options);
  generateFiles(
    tree,
    joinPathFragments(__dirname, 'files/comment'),
    'dist',
    normalizedOptions
  );
  generateFiles(
    tree,
    joinPathFragments(__dirname, 'files/readme'),
    'dist',
    normalizedOptions
  );
}
