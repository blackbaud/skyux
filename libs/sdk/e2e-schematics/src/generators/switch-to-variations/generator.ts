import { Tree, getProjects, logger, visitNotIgnoredFiles } from '@nx/devkit';

import { SwitchToVariationsGeneratorSchema } from './schema';

export default async function (
  tree: Tree,
  options: SwitchToVariationsGeneratorSchema
) {
  const projectConfiguration = getProjects(tree).get(options.project);
  if (!projectConfiguration) {
    throw new Error(`Project ${options.project} not found`);
  }

  // istanbul ignore if
  if (!(projectConfiguration.sourceRoot && projectConfiguration.targets)) {
    throw new Error(`Project source root or targets could not be found`);
  }

  if (!projectConfiguration.targets['e2e']) {
    throw new Error(`Project ${options.project} is not an e2e project`);
  }

  visitNotIgnoredFiles(tree, projectConfiguration.sourceRoot, (filepath) => {
    let reason = '';
    if (filepath.endsWith('.spec.ts') || filepath.endsWith('.cy.ts')) {
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const source = tree.read(filepath, 'utf-8')!;
      // Quick check if this looks like the old pattern
      if (
        source.startsWith(
          `['default', 'modern-light', 'modern-dark'].forEach((theme) => {`
        )
      ) {
        const describePattern =
          /^ {2}describe\(`([-\w ]+) in \$\{theme} theme`, \(\) => \{$/;
        const specLines = source.split(/\r?\n/);
        specLines.shift(); // Remove first line
        const describeLine = `${specLines.shift()}`;
        if (describeLine.match(describePattern)) {
          // Strip empty lines from the bottom of the file
          while (specLines[specLines.length - 1].trim() === '') {
            specLines.pop();
          }
          const lastLine = specLines.pop();
          // Is this a single block?
          if (
            lastLine === '});' &&
            specLines.every((line) => line.startsWith('  ') || line === '')
          ) {
            const newSpecLines = [
              `import { E2eVariations } from '@skyux-sdk/e2e-schematics';`,
              '',
              describeLine.replace(
                describePattern,
                (_, description) => `describe(\`${description}\`, () => {`
              ),
              '  E2eVariations.forEachTheme((theme) => {',
              '    describe(`in ${theme} theme`, () => {',
            ];
            newSpecLines.push(
              ...specLines.map((line) => (line === '' ? '' : `  ${line}`)),
              '  });',
              '});',
              ''
            );
            tree.write(filepath, newSpecLines.join(`\n`));
            return;
          } else if (lastLine === '});') {
            reason = 'multiple blocks';
          } else {
            reason = `last line is ${lastLine}`;
          }
        } else {
          reason = `describe line does not match: ${describeLine}`;
        }
        logger.info(`Unable to update ${filepath}: ${reason}`);
      }
    }
  });
}
