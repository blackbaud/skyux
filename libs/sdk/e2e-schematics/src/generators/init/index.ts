import { storybookConfigurationGenerator } from '@nrwl/angular/generators';
import {
  Tree,
  addDependenciesToPackageJson,
  generateFiles,
  getProjects,
  joinPathFragments,
  logger,
} from '@nrwl/devkit';
import { Linter } from '@nrwl/linter';
import { runTasksInSerial } from '@nrwl/workspace/src/utilities/run-tasks-in-serial';

import { PackageJson } from 'nx/src/utils/package-json';

import {
  applyTransformersToPath,
  getInsertExportTransformer,
} from '../../utils';
import { updateJson } from '../../utils';
import configureStorybook from '../configure-storybook';

import { Schema } from './schema';

/**
 * Verify that Storybook and dependencies are installed.
 */
export default async function (tree: Tree, options: Schema) {
  const storybookConfig = getProjects(tree).get('storybook');
  if (
    !(storybookConfig && storybookConfig.targets && storybookConfig.sourceRoot)
  ) {
    throw new Error(
      'Storybook project or storybook project information not found'
    );
  }
  if (!storybookConfig.targets.storybook) {
    // nx g @nrwl/angular:storybook-configuration storybook --no-configureCypress --no-generateStories --no-generateCypressSpecs
    const storybookInstall = await storybookConfigurationGenerator(tree, {
      name: 'storybook',
      generateStories: false,
      generateCypressSpecs: false,
      configureCypress: false,
      linter: Linter.EsLint,
    });

    // Additional dependencies
    const dependenciesInstall = await addDependenciesToPackageJson(
      tree,
      {},
      {
        '@percy/cli': 'latest',
        '@percy/cypress': 'latest',
        '@percy/storybook': 'latest',
        '@storybook/addon-a11y': 'latest',
        cypress: 'latest',
        'jest-preset-angular': 'latest',
        sb: 'latest',
        'storybook-addon-angular-router': 'latest',
      }
    );

    // Configure storybook composition project
    // nx g @skyux-sdk/e2e-schematics:storybook-configuration storybook
    await configureStorybook(tree, { name: 'storybook' });

    updateJson<PackageJson & { percy?: unknown }>(
      tree,
      'package.json',
      (json) => {
        if (!json.percy) {
          json.percy = {
            version: 2,
            snapshot: {
              widths: [375, 1280],
            },
          };
        }
        return json;
      }
    );

    // Generate Storybook decorators that rely on package dependencies.
    generateFiles(
      tree,
      joinPathFragments(__dirname, './files/preview-wrapper'),
      joinPathFragments(storybookConfig.sourceRoot, './lib/preview-wrapper'),
      {}
    );
    applyTransformersToPath(
      tree,
      joinPathFragments(storybookConfig.sourceRoot, 'index.ts'),
      [
        getInsertExportTransformer(
          './lib/preview-wrapper/preview-wrapper-decorators',
          './lib/storybook.module'
        ),
      ]
    );

    return runTasksInSerial(storybookInstall, dependenciesInstall);
  } else {
    return Promise.resolve(() => {
      (options.ansiColor === false ? console.log : logger.info)(
        'Storybook already initialized'
      );
    });
  }
}
