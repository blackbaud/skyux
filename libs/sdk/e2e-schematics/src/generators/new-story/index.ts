import { componentGenerator } from '@nrwl/angular/generators';
import {
  ProjectConfiguration,
  Tree,
  formatFiles,
  generateFiles,
  getProjects,
  joinPathFragments,
} from '@nrwl/devkit';
import { classify, dasherize } from '@nrwl/workspace/src/utils/strings';

import { angularModuleGenerator } from '../../utils/angular-module-generator';
import { findNgModuleClass, readSourceFile } from '../../utils/ast-utils';
import {
  findClosestModule,
  findModulePaths,
  isRoutingModule,
} from '../../utils/find-module';
import { capitalizeWords, dirname } from '../../utils/utils';
import storiesGenerator from '../stories/index';

import { ComponentGeneratorSchema } from './schema';

interface NormalizedSchema extends ComponentGeneratorSchema {
  projectName: string;
  projectRoot: string;
  projectDirectory: string;
  e2eSourceRoot: string | undefined;
  projectConfig: ProjectConfiguration;
  e2eProjectConfig: ProjectConfiguration;
}

function normalizeOptions(
  tree: Tree,
  options: ComponentGeneratorSchema
): NormalizedSchema {
  const projects = getProjects(tree);
  if (!projects.has(options.project)) {
    throw new Error(`Unable to find project ${options.project}`);
  }
  let projectConfig = projects.get(options.project);
  if (!('storybook' in projectConfig.targets)) {
    options.project = `${options.project}-storybook`;
    if (!projects.has(options.project)) {
      throw new Error(`Unable to find project ${options.project}`);
    }
    projectConfig = projects.get(options.project);
    if (!('storybook' in projectConfig.targets)) {
      throw new Error(`Storybook is not configured for ${options.project}`);
    }
  }
  const projectDirectory = projectConfig.sourceRoot;
  const projectName = options.project;
  const projectRoot = projectConfig.root;
  options.cypressProject = options.cypressProject || `${projectName}-e2e`;
  const e2eProjectConfig = projects.get(options.cypressProject);
  const e2eSourceRoot = e2eProjectConfig.sourceRoot;
  let module: string;
  if (options.module) {
    module = options.module;
  } else {
    const moduleOptions = findModulePaths(tree, projectDirectory, (path) => {
      const sourceFile = readSourceFile(tree, path);
      const module = findNgModuleClass(sourceFile);
      return module && isRoutingModule(module, sourceFile);
    });
    if (moduleOptions.length === 0) {
      throw new Error(
        `Could not find a router module to add the component to. Please specify a module using the --module option.`
      );
    } else if (moduleOptions.length === 1) {
      module = moduleOptions[0]
        .split('/')
        .pop()
        .replace(/\.module\.ts$/, '');
    } else {
      let componentDirectory = `app${
        options.name.includes('/') ? `/${dirname(options.name)}` : ''
      }`;
      module = findClosestModule(
        moduleOptions,
        projectDirectory,
        componentDirectory
      );
    }
  }

  return {
    ...options,
    module,
    projectName,
    projectRoot,
    projectDirectory,
    e2eSourceRoot,
    projectConfig,
    e2eProjectConfig,
  };
}

export default async function (tree: Tree, options: ComponentGeneratorSchema) {
  const normalizedOptions = normalizeOptions(tree, options);
  // Grab previous changes so we can filter them later.
  const previouslyCreated = tree
    .listChanges()
    .filter((change) => change.type === 'CREATE')
    .map((change) => change.path);

  // nx g @schematics/angular:module
  await angularModuleGenerator(tree, {
    name: normalizedOptions.name,
    route: normalizedOptions.name,
    module: normalizedOptions.module,
    project: normalizedOptions.project,
  });
  const componentFilePaths = tree
    .listChanges()
    .filter((change) => change.type === 'CREATE')
    .filter((change) => !previouslyCreated.includes(change.path))
    .filter((change) =>
      change.path.match(/\.component\.(ts|spec\.ts|html|s?css)$/)
    )
    .map((change) => change.path);
  const componentFilePath = componentFilePaths.find((filepath) =>
    filepath.endsWith('.component.ts')
  );
  // Module generator creates the component and adds the route, but it doesn't provide options or use the defaults.
  componentFilePaths.forEach((filepath) => {
    tree.delete(filepath);
  });
  const baseName = normalizedOptions.name.split('/').pop();
  await componentGenerator(tree, {
    name: normalizedOptions.name,
    module: baseName,
    project: normalizedOptions.project,
    style: 'scss',
    displayBlock: true,
    changeDetection: 'OnPush',
    export: true,
    skipImport: true,
  });
  generateFiles(
    tree,
    joinPathFragments(__dirname, 'files'),
    dirname(componentFilePath),
    {
      name: baseName,
      nameDash: dasherize(baseName),
      nameClass: classify(baseName),
      nameCapitalized: capitalizeWords(baseName),
      project: normalizedOptions.project,
      componentPath: dirname(componentFilePath),
      // @ts-ignore
      prefix: normalizedOptions.projectConfig.prefix,
    }
  );

  // nx g @skyux/e2e-schematics:stories
  await storiesGenerator(tree, {
    project: normalizedOptions.project,
    cypressProject: normalizedOptions.cypressProject,
    generateCypressSpecs: normalizedOptions.generateCypressSpecs,
  });

  // Find new files that are not related to the new component and drop them.
  const expectedPaths = [
    `${normalizedOptions.projectDirectory}/app/${normalizedOptions.name}/`,
    `${normalizedOptions.e2eSourceRoot}/integration/${componentFilePath
      .split('/')
      .pop()
      .replace(/\.ts$/, '.spec.ts')}`,
  ];
  const changes = tree.listChanges();
  changes
    .filter((change) => change.type === 'CREATE')
    .filter((change) => !previouslyCreated.includes(change.path))
    .forEach((change) => {
      if (
        expectedPaths.findIndex((expectedPath) =>
          change.path.startsWith(expectedPath)
        ) === -1
      ) {
        tree.delete(change.path);
      }
    });

  await formatFiles(tree);
}
