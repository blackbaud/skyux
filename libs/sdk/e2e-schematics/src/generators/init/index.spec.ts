import { libraryGenerator } from '@nrwl/angular/generators';
import { UnitTestRunner } from '@nrwl/angular/src/utils/test-runners';
import {
  Tree,
  joinPathFragments,
  readProjectConfiguration,
} from '@nrwl/devkit';
import { createTreeWithEmptyWorkspace } from '@nrwl/devkit/testing';
import { Linter } from '@nrwl/linter';

import generator from './index';

describe('init generator', () => {
  let appTree: Tree;

  beforeEach(async () => {
    appTree = createTreeWithEmptyWorkspace();
  });

  it('should run successfully', async () => {
    await libraryGenerator(appTree, {
      name: 'storybook',
      routing: false,
      unitTestRunner: UnitTestRunner.None,
      linter: Linter.None,
    });
    await generator(appTree, { ansiColor: false });
    const config = readProjectConfiguration(appTree, 'storybook');
    expect(config.targets?.storybook).toBeDefined();
    expect(
      appTree.isFile(
        joinPathFragments(
          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
          config.sourceRoot!,
          'lib/preview-wrapper/preview-wrapper-decorators.ts'
        )
      )
    ).toBeTruthy();
    expect(
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      appTree.read(joinPathFragments(config.sourceRoot!, 'index.ts'), 'utf-8')
    ).toMatchSnapshot();
  });

  it('should error without a "storybook" project', async () => {
    try {
      await generator(appTree, { ansiColor: false });
      fail('should have thrown');
    } catch (e) {
      if (!(e instanceof Error)) {
        fail('should have thrown an error');
      }
      expect(e.message).toBe(
        'Storybook project or storybook project information not found'
      );
    }
  });

  it('should skip on subsequent runs', async () => {
    await libraryGenerator(appTree, {
      name: 'storybook',
      routing: false,
      unitTestRunner: UnitTestRunner.None,
      linter: Linter.None,
    });
    const consoleSpy = jest.spyOn(console, 'log');
    await generator(appTree, { ansiColor: false });
    expect(appTree.isFile('libs/storybook/.storybook/preview.ts')).toBeTruthy();
    expect(consoleSpy).not.toHaveBeenCalled();
    let installTask = await generator(appTree, {});
    installTask();
    installTask = await generator(appTree, { ansiColor: false });
    installTask();
    expect(consoleSpy).toHaveBeenCalledWith(`Storybook already initialized`);
  });
});
