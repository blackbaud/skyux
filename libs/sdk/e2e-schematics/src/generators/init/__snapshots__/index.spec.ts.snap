// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`init generator should run successfully 1`] = `
"export * from './lib/storybook.module';
export * from './lib/preview-wrapper/preview-wrapper-decorators';
"
`;
